User and room integration

WhiteDiamondz opened the pull request:
WIP

WhiteDiamondz committed:
Add connectUserToRoom and DisconnectUserFromRoom + extra

Brawdunoir commented on models/room.go:
```diff
@@ -1,19 +1,33 @@
+	UpdatedAt time.Time    `json:"-"`
+	DeletedAt sql.NullTime `gorm:"index" json:"-"`
+	Name      string       `json:"name" binding:"required,gte=2,lte=20" example:"BirthdayParty"`
+	OwnerID   uint         `json:"owner,omitempty"`
+	UsersID   []uint       `json:"users,omitempty"`
```
```suggestion
	OwnerID   uint         `json:"ownerID,omitempty"`
	UsersID   []uint       `json:"usersID,omitempty"`
```

Brawdunoir commented on models/room.go:
```diff
@@ -1,19 +1,33 @@
 
 type RoomUpdate struct {
-	Name string `json:"name,omitempty" binding:"gte=2,lte=20"`
+	ID         uint   `gorm:"primarykey" json:"-"`
+	Name       string `json:"name,omitempty" binding:"gte=2,lte=20"`
+	OwnerID    uint   `json:"owner,omitempty"`
+	NewOwnerID uint   `json:"new_owner"`
+	UsersID    []uint `json:"users,omitempty"`
```
```suggestion
	Name       string `json:"name,omitempty" binding:"gte=2,lte=20"`
	OwnerID    uint   `json:"ownerID,omitempty"`
	UsersID    []uint `json:"usersID,omitempty"`
```
No one should be able to change ID.
Because it is a patch, NewOwnerID is for sure OwnerID.

Brawdunoir commented on routes/room.go:
```diff
@@ -75,7 +86,15 @@ func UpdateRoom(c *gin.Context) {
 	if err != nil {
 		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid room ID"})
-		log.Printf("Failed to convert room ID: %v", err)
+		log.Printf("Failed to extract user from context: %v", err)
```
```suggestion
		log.Printf("Failed to convert room ID: %v", err)
```

Brawdunoir commented on routes/room.go:
```diff
@@ -96,40 +115,109 @@ func UpdateRoom(c *gin.Context) {
 			return
 		}
+	} else {
+		// Check if requester is the owner of the room
+		if user.ID != patchedRoom.OwnerID {
+			c.JSON(http.StatusUnauthorized, gin.H{"error": "You are not authorized to update this room"})
+			log.Printf("Failed to extract user from context: %v", err)
+			return
+		}
```
Use utils.AssertUser() function instead and modify log message

Brawdunoir commented on routes/room.go:
```diff
@@ -96,40 +115,109 @@ func UpdateRoom(c *gin.Context) {
+		return
+	}
+
+	roomID := c.Param("id")
```
Assert ID is a integer
```go
id, err := strconv.Atoi(c.Param("id"))
if err != nil {
…
}
```

Brawdunoir commented on routes/room.go:
```diff
@@ -96,40 +115,109 @@ func UpdateRoom(c *gin.Context) {
 	}
 
-	result := db.WithContext(ctx).Delete(&models.Room{}, id)
+	roomID := c.Param("id")
```
Assert ID is a integer
```go
id, err := strconv.Atoi(c.Param("id"))
if err != nil {
…
}
```

Brawdunoir commented on routes/room.go:
```diff
@@ -96,40 +115,109 @@ func UpdateRoom(c *gin.Context) {
 	}
 
-	c.JSON(http.StatusNoContent, nil)
+	roomUpdate.UsersID = utils.RemoveUintFromSlice(roomUpdate.UsersID, user.ID)
```
What if the user is not in the room ?
Moreover, it seems this function removes the integer at position `user.ID`, not the integer `user.ID`

WhiteDiamondz replied:
Should we really check if the user is connected to the room ? This is triggered by the DisconnectUserFromRoom function where the user we check is the one who sent the request and the room id is passed in the path. 
Better be safe than sorry of course. 
A fix is underway for the remove function

WhiteDiamondz replied:
Has been taken into account:
```
	// Remove user from the connected users list of the room
	roomUpdate.UsersID, err = utils.RemoveUintFromSlice(roomUpdate.UsersID, user.ID)
	if err != nil {
		log.Printf("User not connected to room: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to remove user from room"})
		return
	}
```

Brawdunoir commented on routes/room.go:
```diff
@@ -96,40 +115,109 @@ func UpdateRoom(c *gin.Context) {
+	if room.OwnerID == user.ID {
+		if len(roomUpdate.UsersID) > 0 {
+			// We already removed the user from the slice, so we can get the first element as new Owner
+			roomUpdate.OwnerID = roomUpdate.UsersID[0]
+			err = db.WithContext(ctx).Model(&room).Updates(roomUpdate.ToRoom()).Error
+			if err != nil {
+				log.Printf("Failed to modify document: %v", err)
+				c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to modify document"})
+				return
+			}
```
Externalize this behavior in another function. Will be useful to pass ownership of the room through another route later.

Brawdunoir commented on utils/routes/room.go:
```diff
@@ -0,0 +1,22 @@
+)
+
+// DeleteRoom deletes a room in the database
+func DeleteRoom(id uint, db *gorm.DB) bool {
```
Should return `error` instead of `bool`, according to golang good practises.

WhiteDiamondz committed:
Update models/room.go

Co-authored-by: Yann L. <56452324+Brawdunoir@users.noreply.github.com>

WhiteDiamondz committed:
Update routes/room.go

Co-authored-by: Yann L. <56452324+Brawdunoir@users.noreply.github.com>

WhiteDiamondz committed:
Update models/room.go

Co-authored-by: Yann L. <56452324+Brawdunoir@users.noreply.github.com>

WhiteDiamondz committed:
Fix following first review

WhiteDiamondz committed:
Add basic test

WhiteDiamondz committed:
Remove DeleteRoom test

Brawdunoir committed:
feat: Add pgadmin to the dev docker-compose

Brawdunoir committed:
fix(Model): Add gorm tag to specify type of UsersID field

Brawdunoir committed:
refactor(Connect/Disconnect room): Refactor ConnectUserToRoom/DisconnectUserToRoom

Brawdunoir committed:
chore(PQ): Add pq to packages, it is a postgres driver

We need to add it to use the Int64Array.

It is not possible to handle
arrays otherwise.

Brawdunoir committed:
fix(Rooms type field): Change type of UsersID

Brawdunoir committed:
fix(Router): Use redis cache only if defined previously

Brawdunoir committed:
refactor(Room fields): Remove useless fields

We don’t need these fields anymore as they are not passed in body
requests

Brawdunoir committed:
chore(Package): Add slices experimental package

It provides the contains method since go 1.18

Brawdunoir committed:
feat(Room): Assert the user is not already in the room

Brawdunoir committed:
fix(StatusCode): Change to code 400 when user not in room

Brawdunoir committed:
test(Room): Add tests for Room according to PR

WhiteDiamondz commented on routes/room.go:
```diff
@@ -96,38 +109,133 @@ func UpdateRoom(c *gin.Context) {
+		// Check if requester is the owner of the room
+		err := utilsRoutes.AssertUser(c, int(patchedRoom.OwnerID))
+		if err != nil {
+			log.Printf("Error when asserting user: %v", err)
```
```suggestion
                        c.JSON(http.StatusInternalServerError, gin.H{"error": "Room not modified"})
			log.Printf("Error when asserting user: %v", err)
```

WhiteDiamondz commented on tests/room_test.go:
```diff
@@ -119,40 +124,133 @@ func TestUpdateRoom(t *testing.T) {
+	test.Run(t)
+}
+
+// TestDisconnectRoom tests the DisconnectUserToRoom function.
```
```suggestion
// TestDisconnectRoom tests the DisconnectUserFromRoom function.
```

WhiteDiamondz commented on tests/room_test.go:
```diff
@@ -119,40 +124,133 @@ func TestUpdateRoom(t *testing.T) {
+		CreateRequestHeaders: headers,
+		CreateResponse:       CreateResponseRoom{},
+		SubTests: []utils.SubTest{
+			{Name: "Assert Setup", Request: utils.Request{Method: http.MethodGet, Headers: headers}, ResponseCode: http.StatusOK, ResponseBodyRegex: regex},
```
I don't think this should be necessary. Indirectly it tests the creation of the room and the joining. If we do think it is good practice then we should consider adding the same kind of tests to functions like TestDeleteUser.

vazanoir commented on routes/room.go:
```diff
@@ -96,38 +109,133 @@ func UpdateRoom(c *gin.Context) {
+			log.Printf("Failed to delete document: %v", err)
+			c.JSON(http.StatusInternalServerError, gin.H{"error": "Room not deleted"})
+			return
+		} else if result.RowsAffected < 1 {
+			log.Printf("Failed to find document: %v", err)
+			c.JSON(http.StatusNotFound, gin.H{"error": "Room not found"})
+			return
+		}
```
Not necessary as line 201 get the room and return if not found

vazanoir commented on utils/functions.go:
```diff
@@ -0,0 +1,17 @@
+	"github.com/lib/pq"
+)
+
+func RemoveUintFromSlice(slice pq.Int64Array, value uint) (pq.Int64Array, error) {
```
Couldn't value be uint64 directly ?

vazanoir commented on routes/room.go:
```diff
@@ -96,40 +109,159 @@ func UpdateRoom(c *gin.Context) {
+	// Check if user is owner to know if we give ownership to a new user
+	if room.OwnerID == user.ID {
+		if len(roomUpdate.UsersID) > 0 {
+			err = utilsRoutes.ChangeOwner(roomUpdate, db)
```
Either the function isn't necessary or the error check isn't, right now it's a doublon

vazanoir commented on routes/room.go:
```diff
@@ -96,40 +109,159 @@ func UpdateRoom(c *gin.Context) {
+			}
+
+		} else {
+			err = utilsRoutes.DeleteRoom(room.ID, db)
```
Either the function isn't necessary or the error check isn't, right now it's a doublon.
Also, we know the room exist from line 201, doesn't need to check for that.

Brawdunoir committed:
refactor(PR Review): Refactor code to apply PR code review

vazanoir commented on routes/router.go:
```diff
@@ -63,11 +64,13 @@ func SetupRouter(router *gin.Engine) *gin.Engine {
 				roomRouter.PATCH("/:id", invalidateCacheURI, UpdateRoom)
-				roomRouter.DELETE("/:id", invalidateCacheURI, DeleteRoom)
+				roomRouter.POST("/:id/connect", invalidateCacheURI, ConnectUserToRoom)
+				roomRouter.POST("/:id/disconnect", invalidateCacheURI, DisconnectUserFromRoom)
 			} else {
 				roomRouter.GET("/:id", GetRoom)
 				roomRouter.PATCH("/:id", UpdateRoom)
-				roomRouter.DELETE("/:id", DeleteRoom)
+				roomRouter.POST("/:id/connect", ConnectUserToRoom)
+				roomRouter.POST("/:id/disconnect", DisconnectUserFromRoom)
```
Should be PATCH

vazanoir commented on routes/room.go:
```diff
@@ -96,38 +108,129 @@ func UpdateRoom(c *gin.Context) {
+		c.JSON(http.StatusNoContent, nil)
+		return
+	} else if room.OwnerID == user.ID {
+		room.OwnerID = uint(room.UsersID[0])
```
Unify  all id types

Brawdunoir replied:
Will look into this

Brawdunoir committed:
fix(Room user list): Fix delete in slice upon user disconnect

Brawdunoir committed:
feat(POST to PATCH): Patch seems to be better for connect/disconnect
